### ISSUE TITLE
Multiple orders/fulfils. No MerchantID

### ISSUE THREAD

**Issue by BarryCarlyon** [beardon/mws-api#83]
As per notes in pr #75 and some in #82 

The tests are wrong, the tests omit the required `MerchantIdentifier` and Amazon rejects feeds that don't contain this header key.

XSD is here: https://images-na.ssl-images-amazon.com/images/G/01/rainier/help/xsd/release_1_9/amzn-header.xsd

Currently my test fix is:

Using the merchantID from the first entry in the data. Which then means setting the merchantID on (first/each) passed in message.

Does it make more sense to set the merchantID when constructing the `AmazonMwsClient`

```
const _POST_ORDER_FULFILLMENT_DATA_ = createEnvelope(function (data) {
    const messages = Array.isArray(data) ? data : [data];

    return {
        MerchantIdentifier: messages[0].MerchantIdentifier,
        MessageType: 'OrderFulfillment',
        PurgeAndReplace: messages[0].PurgeAndReplace,
        Message: messages.map((m) => ({
            MessageID: m.MessageID,
            OrderFulfillment: {
                AmazonOrderID: m.AmazonOrderID,
                MerchantOrderID: m.MerchantOrderID,
                MerchantFulfillmentID: m.MerchantFulfillmentID,
                FulfillmentDate: m.FulfillmentDate,
                FulfillmentData: {
                    CarrierCode: m.CarrierCode,
                    CarrierName: m.CarrierName,
                    ShippingMethod: m.ShippingMethod,
                    ShipperTrackingNumber: m.ShipperTrackingNumber
                },
                Item: {
                    MerchantFulfillmentItemID: m.MerchantFulfillmentItemID,
                    Quantity: m.Quantity
                }
            }
        }))
    };
});
```

**Comment by BarryCarlyon** [comment]
Current Test:

```
            it('should handle a single tracking number', () => {
                const FeedContent = Feeds._POST_ORDER_FULFILLMENT_DATA_({
                    MessageID: 1,
                    MerchantOrderID: 1234567,
                    MerchantFulfillmentID: 1234567,
                    FulfillmentDate: '2002-05-01T15:36:33-08:00',
                    CarrierCode: 'UPS',
                    ShippingMethod: 'Second Day',
                    ShipperTrackingNumber: '1234567890'
                });
                const xml = '<?xml version="1.0" encoding="ISO-8859-1"?><AmazonEnvelope xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="amzn-envelope.xsd"><Header><DocumentVersion>1.01</DocumentVersion></Header><MessageType>OrderFulfillment</MessageType><Message><MessageID>1</MessageID><OrderFulfillment><MerchantOrderID>1234567</MerchantOrderID><MerchantFulfillmentID>1234567</MerchantFulfillmentID><FulfillmentDate>2002-05-01T15:36:33-08:00</FulfillmentDate><FulfillmentData><CarrierCode>UPS</CarrierCode><ShippingMethod>Second Day</ShippingMethod><ShipperTrackingNumber>1234567890</ShipperTrackingNumber></FulfillmentData></OrderFulfillment></Message></AmazonEnvelope>'

                expect(FeedContent).to.equal(xml);
            });
```

Replace test with: (then fix code to match)

```
            it('should handle a single tracking number', () => {
                const FeedContent = Feeds._POST_ORDER_FULFILLMENT_DATA_({
                    MessageID: 1,
                    MerchantOrderID: 1234567,
                    MerchantFulfillmentID: 1234567,
                    FulfillmentDate: '2002-05-01T15:36:33-08:00',
                    CarrierCode: 'UPS',
                    ShippingMethod: 'Second Day',
                    ShipperTrackingNumber: '1234567890'
                });
                const xml = '<?xml version="1.0" encoding="ISO-8859-1"?><AmazonEnvelope xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="amzn-envelope.xsd"><Header><DocumentVersion>1.01</DocumentVersion><MerchantIdentifier>67890</MerchantIdentifier></Header><MessageType>OrderFulfillment</MessageType><Message><MessageID>1</MessageID><OrderFulfillment><MerchantOrderID>1234567</MerchantOrderID><MerchantFulfillmentID>1234567</MerchantFulfillmentID><FulfillmentDate>2002-05-01T15:36:33-08:00</FulfillmentDate><FulfillmentData><CarrierCode>UPS</CarrierCode><ShippingMethod>Second Day</ShippingMethod><ShipperTrackingNumber>1234567890</ShipperTrackingNumber></FulfillmentData></OrderFulfillment></Message></AmazonEnvelope>'

                expect(FeedContent).to.equal(xml);
            });
```

for example?

**Comment by BarryCarlyon** [comment]
I'd propose changing:

```
function AmazonEnvelope(data) {
    return {
        '@': {
            'xmlns:xsi': 'http://www.w3.org/2001/XMLSchema-instance',
            'xsi:noNamespaceSchemaLocation': 'amzn-envelope.xsd'
        },
        Header: {
            DocumentVersion: 1.01,
            MerchantIdentifier: data.MerchantIdentifier
        },
        MessageType: data.MessageType,
        Message: data.Message
    };
}
```

in xml.js

the data.MerchantIdentifier is wrong. Since no data get's passed from the multi's

This can be loaded from the constructor but not sure the right way to pass it

**Comment by BarryCarlyon** [comment]
Yeah for this PR ( #82 ) I copied the code for the other multi job and adjusted it, did the branches and made the PR then actually threw a live packet at amazon and got:

```{ result:
   { DocumentTransactionID: '--',
     StatusCode: 'Complete',
     ProcessingSummary:
      { MessagesProcessed: '0',
        MessagesSuccessful: '0',
        MessagesWithError: '1',
        MessagesWithWarning: '0' },
     Result:
      { MessageID: '--',
        ResultCode: 'Error',
        ResultMessageCode: '50001',
        ResultDescription: 'The XML you submitted is ill-formed at the Amazon Envelope XML level at (or near) line 1, column 220..' } },
```

Or various column offsets as I test

Col 220 lines up with the closing `</header>` in:

```
<?xml version="1.0" encoding="ISO-8859-1"?><AmazonEnvelope xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="amzn-envelope.xsd"><Header><DocumentVersion>1.01</DocumentVersion></Header><MessageType>OrderAcknowledgement</MessageType><Message><MessageID>1</MessageID><OrderAcknowledgement><AmazonOrderID>123-0000000-0000000</AmazonOrderID><MerchantOrderID>0000000000</MerchantOrderID><StatusCode>Success</StatusCode></OrderAcknowledgement></Message></AmazonEnvelope>
```

Secrets/sensitives replaced. But there is no `<Header>` key of `<MerchantIdentifier>`

**Comment by BarryCarlyon** [comment]
This issue also affects `PurgeAndReplace` too

**Comment by BarryCarlyon** [comment]
purgeandreplace is not a required field for fulfil or ack but it's in the module code.

I'm not sure if making feeds extend off a client is better, or passing in a header to the feed content creator is better


```
                const FeedContent = Feeds._POST_ORDER_ACKNOWLEDGEMENT_DATA_({MerchantIdentifier: 67890}, [{
                    MessageID: 1,
                    AmazonOrderID: '123-4567890-1234567',
                    MerchantOrderID: '12345',
                    StatusCode: 'Success',
                    MerchantIdentifier: '67890',
                    Item: {
                        AmazonOrderItemCode: '12345',
                        MerchantOrderItemID: '54321'
                    }
                }, {
                    MessageID: 2,
                    AmazonOrderID: '987-6543210-9876543',
                    MerchantOrderID: '98765',
                    StatusCode: 'Success',
                    MerchantIdentifier: '67890',
                    Item: {
                        AmazonOrderItemCode: '54321',
                        MerchantOrderItemID: '12345'
                    }
                }]);
```

 for example

But I think "feed content maker" should extend the base client.

**Comment by dciccale** [comment]
It definitely gives more flexibility allowing passing a header.

However we could check arguments length, if only 1 arg is passed, we treat it as usual. If not, we take 1st arg as header parameters. wdyt?

**Comment by BarryCarlyon** [comment]
After some thought and testing, since the ONLY field that goes in the header is MerchantIdentifier, just pass the constructed client over the stack instead.

#82 contains the fixes to the test and the actual fix that meets the test case.

I tried a few different variations, I think passing the client in, rather than a custom object/header works best.

Another alternative would be to adjust the `SubmitFeed` function.

So a call to `Feeds._POST_ORDER_ACKNOWLEDGEMENT_DATA_({messages})` would only generate`MessageType` and `Message` and `SubmitFeed` would itself inject the `Header` since `SubmitFeed` is a child of MWSClient

**Comment by BarryCarlyon** [comment]
And/or Should we remove MessageID from being user generated. And handle the MessageID counter internally.

Probably not, since MessageID's can be used to processed errors, so the User will want to set'em

But we could check for a MessageID and determine it if missing?

**Comment by BarryCarlyon** [comment]
```

<?xml version="1.0" encoding="UTF-8"?>
<AmazonEnvelope xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="amzn-envelope.xsd">
	<Header>
		<DocumentVersion>1.02</DocumentVersion>
		<MerchantIdentifier>Removed</MerchantIdentifier>
	</Header>
	<MessageType>ProcessingReport</MessageType>
	<Message>
		<MessageID>1</MessageID>
		<ProcessingReport>
			<DocumentTransactionID>removed</DocumentTransactionID>
			<StatusCode>Complete</StatusCode>
			<ProcessingSummary>
				<MessagesProcessed>1</MessagesProcessed>
				<MessagesSuccessful>1</MessagesSuccessful>
				<MessagesWithError>0</MessagesWithError>
				<MessagesWithWarning>0</MessagesWithWarning>
			</ProcessingSummary>
		</ProcessingReport>
	</Message>

```

The fix worked \o/

**Comment by dciccale** [comment]
I understand the approach you proposed.

I am not really convinced on having to pass an instance of a class to each method which in the end it just generates a string of XML.

I am more into this kind of option

```js
Feeds._POST_ORDER_ACKNOWLEDGEMENT_DATA_({
    MerchantIdentifier: '67890'
  }, {
    MessageID: 1,
    AmazonOrderID: '123-4567890-1234567',
    MerchantOrderID: '12345',
    StatusCode: 'Success'
});
```

Even if it looks verbose. I think is more testable, and using plain objects.

In a real scenario (such as in my case), I am anyway creating an instance of mws-api with the merchantId, and I am anyway passing it down to the Feeds. However in the previous version was being added into the now {messages} object.. but since now this can be an array, MerchantIdentifier is not actually part of a message. But of the header. So for me it makes sense to kind of keep this kind of structure when calling the xml generator function passing this 1st optional argument with MerchantIdentifier in it as the header. This would require no deps for testing those functions.

I have been looking at the other option you propose, by injecting the header in SubmitFeed which is already a child from MWSClient and thus having access by default to the `merchantId` prop. This option would be killer. Though for testing feeds alone, we'll anyway have to somehow pass the merchantId to the feed functions like `_POST_ORDER_ACKNOWLEDGEMENT_DATA_` in order to get a full testable XML.

**Comment by BarryCarlyon** [comment]
Yeah I think the header injection is the better option, but then we do end up generation xml/parts of XML in two places.

I went with passing a instance of mws-api to save having to global the config since the client is globalled to use anyway.

What do the other contributors think?


I'm running something along the lines of:

```
var amazonconfig = JSON.parse(fs.readFileSync(__dirname + '/config.json'));

const MWSClient = require('mws-api');
Feeds = require('mws-api/feeds');

/* construct client */
mws = new MWSClient({
    accessKeyId:        amazonconfig.access_key_id,
    secretAccessKey:    amazonconfig.secret_access_key,
    merchantId:         amazonconfig.merchant_id,
    host:               'mws.amazonservices.co.uk',
    meta: {
        retry: true, // retry requests when throttled
        next: true, // auto-paginate
        limit: Infinity // only get this number of items (NOT the same as MaxRequestsPerPage)
    }
});

// inc
require('./functions/orders')();
require('./functions/ack')();
```

Then my code in orders and ack only have to worry about calling `mws` and `Feeds` rather than `mws`, `Feeds` and `amazonconfig`

**Comment by dciccale** [comment]
Having globals though it is considered a bad practice.. you can't expect having the same value after requiring a node module since it could've been overwritten by accident making the whole thing pretty unstable.
For now it seems that it is more reliable using the instance of mws. Doing a more composable approach would require some refactor

**Comment by BarryCarlyon** [comment]
Yeah I should really change the: 

```
require('./functions/orders')();
```

to

```
require('./functions/orders')({
mws: mws,
feed: feed
});
```

Or something but thats how I have my implmenetation right now

**Comment by BarryCarlyon** [comment]
Closing issue as discussion and fixes are superseded by #82
